Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
 
   switch (tag) {
     case google_firestore_v1beta1_Value_null_value_tag:
-      DecodeNull(&stream);
+      DecodeNull(stream);
       return FieldValue::NullValue();
     case google_firestore_v1beta1_Value_boolean_value_tag:
-      return FieldValue::BooleanValue(DecodeBool(&stream));
+      return FieldValue::BooleanValue(DecodeBool(stream));
     case google_firestore_v1beta1_Value_integer_value_tag:
-      return FieldValue::IntegerValue(DecodeInteger(&stream));
+      return FieldValue::IntegerValue(DecodeInteger(stream));
     case google_firestore_v1beta1_Value_string_value_tag:
-      return FieldValue::StringValue(DecodeString(&stream));
+      return FieldValue::StringValue(DecodeString(stream));
+    case google_firestore_v1beta1_Value_map_value_tag:
+      return FieldValue::ObjectValue(DecodeObject(stream));
 
     default:
       // TODO(rsgowman): figure out error handling
       abort();
   }
 }
 
+void EncodeSubFieldValue(pb_ostream_t* stream, const FieldValue& field_value) {
+  // Implementation note: This is roughly modeled on pb_encode_delimited (which
+  // is actually pb_encode_submessage), adjusted to account for the oneof in
+  // FieldValue.
+
+  // First calculate the message size using a non-writing substream.
+  pb_ostream_t substream = PB_OSTREAM_SIZING;
+  EncodeFieldValueImpl(&substream, field_value);
+  size_t size = substream.bytes_written;
+
+  // Write out the size to the output stream.
+  EncodeVarint(stream, size);
+
+  // If stream is itself a sizing stream, then we don't need to actually parse
+  // field_value a second time; just update the bytes_written via a call to
+  // pb_write. (If we try to write the contents into a sizing stream, it'll
+  // fail since sizing streams don't actually have any buffer space.)
+  if (stream->callback == NULL) {
+    bool status = pb_write(stream, NULL, size);
+    if (!status) {
+      // TODO(rsgowman): figure out error handling
+      abort();
+    }
+    return;
+  }
+
+  // Ensure the output stream has enough space
+  if (stream->bytes_written + size > stream->max_size) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+
+  // Use a substream to verify that a callback doesn't write more than what it
+  // did the first time. (Use an initializer rather than setting fields
+  // individually like nanopb does. This gives us a *chance* of noticing if
```

Comment:
I think initializer is cleaner, but I'm not sure how it could help with new fields being added. I think this will still compile if a new member were added to the struct after the `errmsg`.

Reply:
I don't think it will. I tested by removing the errmsg=Null field (which causes a compilation error.)